subsys/bluetooth/host/keys.c

```diff
@@ -509,3 +509,16 @@ void bt_keys_show_sniffer_info(struct bt_keys *keys, void *data)
 	}
 }
 #endif /* defined(CONFIG_BT_LOG_SNIFFER_INFO) */
+
+#ifdef ZTEST_UNITTEST
```

Comment:
As discussed previously, there are definite concerns about this approach to white-box testing, especially in the context of a security-related module like the key store. Generally, introducing additional interfaces to increase testing coverage of a module is a good thing except for some side effects. The most subtle, but still risky is the potential variation in behavior, especially failure modes, that may occur from having the code that is tested vary from the code that runs. I would be surprised if introducing these methods would ever result in a failure ocurring in the real code but not under test, but it's not impossible. If a failure were to occur it would be from an erroneous or error-prone use of function pointers, offsets, hand-rolled setjmp/longjmp behavior where the presence or absence of code changes behavior. I would consider this a vanishingly small probability, but it is not without precedent. It is also a problem because if it works in unit testing but fails in practice, that leaves a bug, potentially a very difficult to find bug.

The second issue is that of encapsulation violation. This is not just of academic interest. If the bt_keys structure is modified or eliminated in lieu of a different/better approach without other changes to the interface, the unit tests become a dependency that breaks despite no other breakage. More importantly, this invites a future developer to assume access to the key pool is okay and that the interface can be accessed and manipulated simply by enabling the unit-test macro. This could result in introduced dependencies that are hard to eliminate. Again, this is a future-proofing concern and can be worked around, but it leads to the third concern:

The data in the key pool is security-related and any access to the location in memory of the data through an application-level interface opens up more avenues for manipulation and possible future attack vectors. Admittedly, the existing interfaces does not appear hardened against manipulation, but this is another level of access that could be abused. This is less relevant for other interfaces, but an invitation to make 'tricky' fixes is a problem in code designed to be maintained.

Generally if a white-box testing approach is being used and metadata needs to be accessed to show that internal behavior operated correctly, such interfaces changes need to be public and permanent (which may be a hard sell for this and similar modules). In many cases, black-box testing (testing the interface and external behavior without consideration of the internal states) is sufficient, but if the intention is to introduce full coverage testing, that may require specific new behaviors.

I would caution that changes like this (if not already present in the code) have a good chance of being rejected or causing a flame war. We should discuss Nordic's needs and suggest alternatives that are more likely to be approved (or test the waters and find out before introducing a lot of new dependencies).

Reply:
We can add that concern on top of items list to be discussed with Nodric team